src/sinks/vector/v2.rs

```diff
@@ -0,0 +1,339 @@
+
+    if let Ok(response) = request.await {
+        let status = proto::ServingStatus::from_i32(response.into_inner().status);
+
+        if let Some(proto::ServingStatus::Serving) = status {
+            return Ok(());
+        }
+    }
+
+    Err(Box::new(Error::Health))
+}
+
+fn get_authority(url: &str) -> Result<String, Error> {
+    url.parse::<Uri>()
+        .ok()
+        .and_then(|uri| uri.authority().map(ToString::to_string))
+        .ok_or(Error::NoHost)
+}
+
+impl tower::Service<Vec<proto::EventRequest>> for Client {
+    type Response = ();
+    type Error = Error;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // Readiness check of the client is done through the `push_events()`
+        // call happening inside `call()`. That check blocks until the client is
+        // ready to perform another request.
+        //
+        // See: <https://docs.rs/tonic/0.4.2/tonic/client/struct.Grpc.html#method.ready>
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, requests: Vec<proto::EventRequest>) -> Self::Future {
+        let mut futures = Vec::with_capacity(requests.len());
+
+        // TODO: Instead of firing off multiple requests, have the server accept
+        // more than one event per request (i.e. bulk endpoint).
+        for request in requests {
+            let mut client = self.clone();
+            futures.push(async move { client.push_events(request.into_request()).await })
+        }
+
+        Box::pin(async move {
+            future::join_all(futures)
+                .await
+                .into_iter()
+                .try_for_each(|v| match v {
+                    Ok(..) => Ok(()),
+                    Err(err) => Err(Error::Request { source: err }),
+                })
+        })
+    }
+}
+
+fn encode_event(event: Event) -> EncodedEvent<proto::EventRequest> {
+    let request = proto::EventRequest {
+        message: Some(event.into()),
+    };
+
+    EncodedEvent::new(request)
+}
+
+impl EncodedLength for proto::EventRequest {
+    fn encoded_length(&self) -> usize {
+        self.encoded_len()
+    }
+}
+
+impl sink::Response for Response {
+    fn is_successful(&self) -> bool {
+        self.is_ok()
+    }
+}
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Request failed: {}", source))]
+    Request { source: tonic::Status },
+
+    #[snafu(display("Vector source unhealthy"))]
+    Health,
+
+    #[snafu(display("URL has no host."))]
+    NoHost,
+}
+
+#[derive(Debug, Clone)]
+struct VectorGrpcRetryLogic;
+
+impl RetryLogic for VectorGrpcRetryLogic {
+    type Error = Error;
+    type Response = ();
+
+    fn is_retriable_error(&self, err: &Self::Error) -> bool {
+        if let Error::Request { source } = err {
+            if let tonic::Code::Unknown = source.code() {
+                return false;
+            }
+        }
+
+        true
```

Comment:
IMO a match would be cleaner here:
```suggestion
        match err {
            Error::Request { source } => !matches!(source.code(), tonic::Code::Unknown),
            _ => true,
        }
```